### PR TITLE
Input a class of students via an external text file

### PR DESCRIPTION
**Pull request by ParthGandhiNUS**
Issue #82 now solved...

### CODE REVIEW

File: src/main/java/classify/textfilecode/TextFileParser.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+            while (line.ready()){
+                String input = line.readLine();
+                decodeTextFile(input, subjectName, masterStudentList);
+            }
+        } catch (IOException e) {
+            UI.println(ERROR_ACCESSING_THE_FILE);
+        }
+    }
+
+    private static void decodeTextFile (String textFileInput, String subjectName,
```
Comment (Cryolian) [inline src/main/java/classify/textfilecode/TextFileParser.java:123]: A lot of repeated code here from the data package. Perhaps make a function for both for the purposes of reading in a text file?

File: src/main/java/classify/textfilecode/TextFileParser.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        student.getAttributes().setRemarks(inputArr[REMARKS].trim());
+        
+        String [] subjectInformation = inputArr[SUBJECTS].trim().split(SUBJECT_INFO_REGEX);
+        //Set grades
+        double subjectGrades = Double.parseDouble(subjectInformation[GRADE_FOR_SUBJECT].trim());
+        //Set Classes Attended
+        int subjectClassesAttended = Integer.parseInt(subjectInformation[NUMBER_OF_CLASSES_ATTENDED].trim());
+        SubjectGrade newSubject = new SubjectGrade(subjectName, subjectGrades, subjectClassesAttended);
+        student.getAttributes().addSubjectGrade(newSubject);
+    }
```
Comment (Cryolian) [inline src/main/java/classify/textfilecode/TextFileParser.java:171]: Function seems a little too long as well, about 72 lines. Maybe try refactoring some portions out?

File: src/main/java/classify/textfilecode/TextFileParser.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    private static final String ERROR_ACCESSING_THE_FILE = "Something went wrong while accessing the file";
+    private static final Integer NAME = 0;
+    private static final Integer GENDER = 1;
+    private static final Integer PHONE_NUMBER = 2;
+    private static final Integer LAST_PAYMENT_DATE = 3;
+    private static final Integer REMARKS = 4;
+    private static final Integer SUBJECTS = 5;
+    private static final Integer GRADE_FOR_SUBJECT = 0;
+    private static final Integer NUMBER_OF_CLASSES_ATTENDED = 1;
+    private static final String INPUT_TEXT_FILE_DIRECTORY = "./data/inputFolder";
```
Comment (Cryolian) [inline src/main/java/classify/textfilecode/TextFileParser.java:42]: Repeated constants from the data package. Perhaps just make them public in their other package and call them over?

File: src/main/java/classify/textfilecode/TextFileParser.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    private static void decodeTextFile (String textFileInput, String subjectName,
+                                        ArrayList <Student> masterStudentList){
+        if (textFileInput.trim().isBlank()){
+            return;
+        }
+        String [] inputArr = textFileInput.split(MAIN_REGEX);
+
+        //Set Name
+        Student student = new Student(inputArr[NAME].trim());
+        masterStudentList.add(student);
```
Comment (Cryolian) [inline src/main/java/classify/textfilecode/TextFileParser.java:132]: Maybe adding to the database should be done at the end of the function, when no errors regarding the data are created?

**Review by Cryolian — COMMENTED** [review]
While the code looks mostly acceptable, I do not really see the point of this function given the current implementation, especially with the repeated code. It might be less troublesome for the user to just directly edit the save file we already have after exiting the program.

File: src/main/java/classify/user/InputParsing.java [hunk 5]
```diff
@@ -106,6 +108,11 @@ public static void parseUserCommand(String[] userCommand, ArrayList<Student> mas
                 parseListCommand(masterStudentList, archiveList, recentlyDeletedList, in);
             }
             break;
+        
+        case PROCESS_FILE:
```
Comment (blackmirag3) [inline src/main/java/classify/user/InputParsing.java:112]: Valid naming convention for magic literals, standardised with the rest of our programme 👍

File: src/main/java/classify/textfilecode/TextFileParser.java (hunk 3 above)
Comment (blackmirag3) [reply src/main/java/classify/textfilecode/TextFileParser.java:42]: Second this, superficial point but can also consider implementing printing methods in UI class to reuse within data packages

File: src/main/java/classify/textfilecode/TextFileParser.java (hunk 4 above)
Comment (blackmirag3) [reply src/main/java/classify/textfilecode/TextFileParser.java:132]: Maybe could call add student after adding phone number to the current student instance if you want to allow user to read in corrupted but still valid (name and number) students

**Review by blackmirag3 — APPROVED** [review]

**Pull request closed without merging**